### PR TITLE
feat: refactor chaper 0 + integrate features of OZ 0.20.0

### LINKED ISSUE #172: [FEAT] refactor chaper 0 + integrate features of OZ 0.20.0
## Issue Overview

Refactor the `SimpleNFT.cairo` contract to use OpenZeppelin 0.20.0

## Proposed Solutions or Ideas

- Upgrade OpenZeppelin
- Refactor the NFT contract to ensure no errors are happening
- Test them out

### PR DESCRIPTION
**Pull request by manlikeHB**
# Task name here

Fixes #172 

## Types of change

- [ ] Enhancement

### CODE REVIEW

File: .tool-versions [hunk 1]
```diff
@@ -1,3 +1,3 @@
-scarb 2.8.5
+scarb 2.9.1
```
Comment (metalboyrick) [inline .tool-versions:1]: please bump to `2.9.2`

**Review by metalboyrick — CHANGES_REQUESTED** [review]
cc @Nadai2010

**Comment by Nadai2010** [comment]
The main idea here is to make the "[`YourCollectible`](https://github.com/Scaffold-Stark/speedrunstark/blob/challenge-0-simple-nft/packages/snfoundry/contracts/src/YourCollectible.cairo)" NFT better with the best possible integrations of the [latest versions the OZ](https://github.com/OpenZeppelin/cairo-contracts/releases), you can check [the news ERC721 the OZ](https://github.com/Scaffold-Stark/speedrunstark/blob/challenge-0-simple-nft/packages/snfoundry/contracts/src/test/TestContract.cairo).

- Check that the deployment in devnet 0.2.3 and sepolia works well with ch-0 in UI.
- Pass all tests of snfoundry 0.34.0 and adapt the [TestContract.cairo](https://github.com/Scaffold-Stark/speedrunstark/blob/challenge-0-simple-nft/packages/snfoundry/contracts/src/test/TestContract.cairo) for the new contract.

File: .tool-versions [hunk 2]
```diff
@@ -1,3 +1,3 @@
...
 starknet-foundry 0.33.0
```
Comment (Nadai2010) [inline .tool-versions:2]: starknet-foundry 0.34.0
starknet-devnet 0.2.3

File: packages/snfoundry/contracts/Scarb.toml [hunk 3]
```diff
@@ -6,14 +6,16 @@ edition = "2023_11"
...
-openzeppelin_access = "0.19.0"
-openzeppelin_token = "0.19.0"
+openzeppelin_access = "0.20.0"
+openzeppelin_token = "0.20.0"
+openzeppelin_introspection = "0.20.0"
 
 [dev-dependencies]
-openzeppelin_utils = "0.19.0"
+openzeppelin_utils = "0.20.0"
 snforge_std = "0.33.0"
```
Comment (Nadai2010) [inline packages/snfoundry/contracts/Scarb.toml:17]: 0.34.0

File: packages/snfoundry/contracts/Scarb.toml [hunk 4]
```diff
@@ -6,14 +6,16 @@ edition = "2023_11"
 # See more keys and their definitions at https://docs.swmansion.com/scarb/docs/reference/manifest.html
 
 [dependencies]
-starknet = "2.8.5"
+starknet = "2.9.1"
```
Comment (Nadai2010) [inline packages/snfoundry/contracts/Scarb.toml:9]: 2.9.2

**Review by Nadai2010 — CHANGES_REQUESTED** [review]

**Comment by manlikeHB** [comment]
After implementing the requested changes, everything still works as expect both on devnet and sepolia testnet. all tests pass locally as well.

If there is any particular changes you'd like to see let me know please.

**Comment by Nadai2010** [comment]
Hi, thanks for the contribution, although I don't see any changes in YourCollectible.cairo, do you think you can refactor or are there important improvements in the latest versions to integrate into this contract?

It should also be synchronized with the latest version of base-template [this last PR](https://github.com/Scaffold-Stark/speedrunstark/pull/174)

**Comment by manlikeHB** [comment]
There are no improvements to be added, most of the changes don't affect the current contract/implementation. but for synchronization with the base-template, I'll work on that.

**Review by Nadai2010 — COMMENTED** [review]
Perfect, thanks for your contributions.

**Comment by manlikeHB** [comment]
Sorry this took a while, faced some issues while syncing causing the code to break at different points but its all good and fixed now.

File: packages/nextjs/contracts/deployedContracts.ts
Comment (Nadai2010) [inline packages/nextjs/contracts/deployedContracts.ts:1]: This file no commit please

**Review by Nadai2010 — CHANGES_REQUESTED** [review]
Good job @manlikeHB , thank you very much, looking forward to the small change for the merged one

**Pull request merged**